### PR TITLE
Suggest to shorten temporary borrow from raw pointer

### LINKED ISSUE #65436: async/await regression regarding borrows across yields
The following code was working in nightly-2019-09-05, but no longer works in nightly-2019-10-15.  The workaround is not quite intuitive.

```rust
struct Foo(*const u8);

unsafe impl Send for Foo {}

async fn bar(_: Foo) {}

fn assert_send<T: Send>(_: T) {}

fn main() {
    assert_send(async {
        // This works in nightly-2019-09-05, but no longer works in nightly-2019-10-15:
        bar(Foo(std::ptr::null())).await;
        // This way it does work in nightly-2019-10-15:
        //let foo = Foo(std::ptr::null());
        //bar(foo).await;
    })
}

```

([Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=55bdcd46d1236a7a09fcb32d9175a81b))

Errors:

```
   Compiling playground v0.0.1 (/playground)
error[E0277]: `*const u8` cannot be sent between threads safely
  --> src/main.rs:10:5
   |
7  | fn assert_send<T: Send>(_: T) {}
   |    -----------    ---- required by this bound in `assert_send`
...
10 |     assert_send(async {
   |     ^^^^^^^^^^^ `*const u8` cannot be sent between threads safely
   |
   = help: within `impl std::future::Future`, the trait `std::marker::Send` is not implemented for `*const u8`
   = note: required because it appears within the type `{fn(Foo) -> impl std::future::Future {bar}, fn(*const u8) -> Foo {Foo}, fn() -> *const u8 {std::ptr::null::<u8>}, *const u8, Foo, impl std::future::Future, impl std::future::Future, ()}`
   = note: required because it appears within the type `[static generator@src/main.rs:10:23: 16:6 {fn(Foo) -> impl std::future::Future {bar}, fn(*const u8) -> Foo {Foo}, fn() -> *const u8 {std::ptr::null::<u8>}, *const u8, Foo, impl std::future::Future, impl std::future::Future, ()}]`
   = note: required because it appears within the type `std::future::GenFuture<[static generator@src/main.rs:10:23: 16:6 {fn(Foo) -> impl std::future::Future {bar}, fn(*const u8) -> Foo {Foo}, fn() -> *const u8 {std::ptr::null::<u8>}, *const u8, Foo, impl std::future::Future, impl std::future::Future, ()}]>`
   = note: required because it appears within the type `impl std::future::Future`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0277`.
error: could not compile `playground`.

To learn more, run the command again with --verbose.

```

### PR DESCRIPTION
**Pull request by csmoe**
Closes https://github.com/rust-lang/rust/issues/65436
r? @estebank 
cc @tmandry

### CODE REVIEW

File: src/test/ui/async-await/issues/issue-65436-raw-ptr-not-send.rs [hunk 1]
```diff
@@ -0,0 +1,17 @@
+// compile-fail
```
Comment (Centril) [inline src/test/ui/async-await/issues/issue-65436-raw-ptr-not-send.rs]: This directive doesn't exist.

File: src/librustc/traits/error_reporting.rs [hunk 2]
```diff
@@ -2612,14 +2615,40 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
...
+                // ^^^^^^^ -- a temporary `&T` created inside this method call due to `&self`
+                // ```
+                //
+                let is_region_borrow =
+                    tables.expr_adjustments(expr).iter().any(|adj| adj.is_region_borrow());
+
+                // ```rust
+                // struct Foo(*const u8);
+                // bar(Foo(std::ptr::null())).await;
+                //     ^^^^^^^^^^^^^^^^^^^^^ -- raw-ptr `*T` created inside this struct ctor.
```
[the remark this answers is not in the document]
Comment (tmandry) [reply src/librustc/traits/error_reporting.rs]: I think they're right with the `--` removed.

(I read `--` as a "line to the arrow," which I think was the intent. But we use `|` for that within error messages, so it's confusing.)

**Comment by yshui** [comment]
#65436 is about general `!Send` types, yet this PR seems (IIUC) to just cover just raw pointer. Could this PR really close #65436?

i.e. https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=e7538e93bf6ee4a748410a86d21ce0d7

**Comment by estebank** [comment]
r=me after rebase

**Comment by tmandry** [comment]
> #65436 is about general `!Send` types, yet this PR seems (IIUC) to just cover just raw pointer. Could this PR really close #65436?

I agree it would be nice to generalize this (maybe in a future PR) to all `!Send` types. I don't know how difficult that would be, though.

**Comment by csmoe** [comment]
@estbank ping

**Pull request merged**